### PR TITLE
Climsoft dialogue

### LINKED ISSUE #4915: Enhance the Climatic > File > CLIMSOFT dialogue
There is a bug that I hope can be fixed by Verison 4.23.  Then there are various enhancements that would be useful in their own right, and would also set a good example for CLIMSOFT dialogues.

Only the bug need be fixed for Rwanda - other stuff would be nice!
1. The bug is that the connection to the database can only be made once in a session.  It works fine.  But if you try again after disconnection, then it gives an error.  It has a line in the error message I don't quite understand which is something like widget followed by a number.  This number gets one larger every time you try the connection.
2. **The sub-dialogue** (making the connection could now be made easier to use.  There are 4 fields.
The first is the name of the database.  There are 2 names now.  I suggest the first is given as the default, and there is a pull-down, tha currently just contains the 2 names.  But there could be more later I assume. The first (default) is mariadb_climsoft_db_v4 and the second is mariadb_climsoft_test_db_v4.
The third field could perhaps also be a pull-down.  It contains 3308 as the default and I'll check with Marcellin what other values might sensibly be there.  For now also have 3306.
3. The **main dialogue** could be made a bit easier.  It could also add flexibility in what is copied over.  Some aspects need more thought, but: 
a) A simple improvement would be to be able to display the station names as well, or instead of the IDs in the selector.  And possibly, with a tooltip, to be able to display other station information, particularly the start (and possibly the end) of the record at that station.
b) A second could be to use the Element information in the dialogue.  Elements each have a **type**.  They are Daily, Monthly, Dekadal, Hourly, Synop and AWS and All.  If just one type is selected, then only the elements of that type are shown.  Eventaully perhaps only those stations are shown in the selector that has data of that type - but that might be tricky with the current software.
What would be useful from that, is the date-time column that is in the output.  It is currently not a date column, but could become one if the data were daily or monthly.
c)  There could also usefully be a checkbox for Transfer Element table.  Currently the Element information is given in the data table.  It is given in 3 ways, and that is tricky, because when made into factors, those columns will be in different orders of their levels. It would be better to just give the element number abd give the other information in the element table.
d) When multiple elements are transferred there could be a Tidy checkbox.  This does an unstack.
d) Later there might also be an Options button perhaps, as there is in the NetCDF dialogue.  
3. a) The main output data frame would become simpler in just giving the Element numbers.  And this column is made into a factor.
b) The Date column is made into a date when the data are daily or monthly.
c) I could not format the data column (perhaps these small numbers again?
d) Perhaps the data could be unstacked, when multiple elements are output.  I am nogt convinced this will always work - for example if there are duplicate values.  So it should be an option.

### PR DESCRIPTION
**Pull request by Ivanluv**
fixes #4915

### CODE REVIEW

**Comment by maxwellfundi** [comment]
@dannyparsons @rdstern We discussed with @Ivanluv about having the  Unstack Data on climsoft dialog and here is what we have come up with. 

@dannyparsons please have a look at the R code to see if there is anything else we might need to add. 

@rdstern you can also have a look at how this works, it might impress you :-)

**Comment by rdstern** [comment]
I will have a look.  @maxwellfundi I am generally getting impressed at all the progress that is being made.  I want to write to the Climsoft team as well, once we have this next version, so this is all good.

File: instat/static/InstatObject/R/instat_object_R6.R [hunk 1]
```diff
@@ -1574,6 +1574,11 @@ DataBook$set("public", "import_from_climsoft", function(stations = c(), elements
       my_stations = paste0("(", paste(as.character(unique(station_data$recordedFrom)), collapse=", "), ")")
       station_info <-  DBI::dbGetQuery(con, paste0("SELECT * FROM station WHERE stationID in ", my_stations, ";"))
     }
+    
+   
+    if (unstack){
+      station_data <- reshape2::dcast(data=station_data, formula=recordedFrom + obsDatetime ~ abbreviation, value.var="obsValue")
```
Comment (dannyparsons) [inline instat/static/InstatObject/R/instat_object_R6.R]: spaces needed around `=`

This looks good to get the data in a nice format, but we are then losing some of the other columns: `describedBy` and `elementName` which we shouldn't just forget.

With this format we should also get another small data frame of element metadata with one row per element. And these two data frames are linked.
We can get this by running another database query on the `obselement` table and get the columns: `describedBy`, `abbreviation`, `elementName`.
Then these two data frames have the same information as in the stacked data format.

Does this make sense? @Patowhiz could help with constructing the query if you need help.

Comment (Patowhiz) [reply instat/static/InstatObject/R/instat_object_R6.R]: @Ivanluv I'm happy to help with the SQL query. To get a resultset form mysql database you just need to use a select statement. 
The structure goes like 

`SELECT * FROM table_name` 

where the table_name is replaced with the real table name you wish to get the data from.

In our case the query could be something like(Assuming you don't want all the columns from the database table);

`SELECT describedBy, abbreviation, elementName FROM obselement`

**Comment by Ivanluv** [comment]
@Patowhiz  where should this be run exactly within `import_from_climsoft` ?

**Comment by maxwellfundi** [comment]
From Danny's explanation, my understanding was where you are stacking the
data, then you need to run the query meaning you will then have three
dataframes in the end namely station info, data , elements info.
@dannyparsons is this correct?

**Comment by dannyparsons** [comment]
yes exactly, that's my understanding of what we want. @rdstern can confirm. 
There should also be a link created between the data and element info (by the element ID column)

File: instat/static/InstatObject/R/instat_object_R6.R (hunk 1 above)
Comment (Patowhiz) [reply instat/static/InstatObject/R/instat_object_R6.R]: @dannyparsons we are checking on this issue with @Ivanluv , checking on your comment above, about getting the element information from obselement, we would like to point out that obselement doesn't have `describeBy `column. Should we assume you only mean't `abbreviation,elementName` ?

Comment (dannyparsons) [reply instat/static/InstatObject/R/instat_object_R6.R]: Yes, that's right thanks for clarifying. I only meant to indicate the field above which are coming from `obselement`.
Essentially, we want to make sure that whether you stack or not, you get the same information from the database (just in a different form).

**Comment by Ivanluv** [comment]
@dannyparsons  are we to have every column from `obselement`?Is this implementation right?

**Comment by dannyparsons** [comment]
That's a good question. We should have flexibility in choosing which columns to include in general in the dialog.
For now just include the same columns from `obselement` as in the original code. This is something we should look at in further improvements to the dialog.

**Comment by shadrackkibet** [comment]
@Ivanluv what is the state of this? is testing needed?

**Comment by Ivanluv** [comment]
Ive not made any changes on this

**Comment by lloyddewit** [comment]
@Ivanluv Please could you clarify the status? thanks

**Comment by Ivanluv** [comment]
This requires more disscussion

**Comment by Patowhiz** [comment]
@Ivanluv I'm happy to be involved in those discussions. This is one of the dialogs that the ICPAC staff were interested in. And I strongly felt that it could be made much better.

**Comment by Patowhiz** [comment]
@rdstern @Ivanluv what's the state of the discussion on this?

**Comment by Patowhiz** [comment]
@rdstern @Ivanluv I'm also happy to look into this and help solve the remaining core items raised in issue if need be.

**Comment by Patowhiz** [comment]
@Ivanluv based on the discussions I had with @rdstern and @lloyddewit, I'm now looking into the issue. I'll be creating a new PR in regards to the issue and I'm looking into taking some of the changes you have made in this PR. I can see you have spent a considerable amount of time in solving this and thus I'll be consulting you to help me on some of the changes in the new PR. Thank you.

**Comment by lloyddewit** [comment]
@Patowhiz @Ivanluv I was wondering how much of this PR's functionality is covered by  PR #5960 and @Patowhiz 's other Climsoft work. The number of code changes in this PR is actaully quite small. Could @Patowhiz integrate this into his work so that we can close this PR? 
thanks

**Comment by lloyddewit** [comment]
@Ivanluv I see that you closed this PR. We need to ensure that these changes are not overlooked. Please could you add a comment to this PR explaining

- if these changes are still needed
- if so, in which PR you expect the changes to be made
- if you'll make the changes yourself, or if you've coordinated with someone else to make the changes

Thanks

**Comment by Ivanluv** [comment]
@lloyddewit the changes I made are no longer needed .@Patowhiz took over this task

**Comment by Patowhiz** [comment]
@Ivanluv thank you for clarifying.

**Comment by lloyddewit** [comment]
Understood, thanks for the clarifications :)

**Pull request closed without merging**